lte/gateway/python/magma/kernsnoopd/ebpf/packet_count.bpf.c

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2020 The Magma Authors.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This program attaches to the packet transmit event net_dev_start_xmit. It
+ * counts the number of bytes and packets sent by each linux binary.
+ * Expect considerable performance impact.
+ * https://patchwork.ozlabs.org/patch/309437/
+ */
+
+// Create a hash map with `key_t` as key type and `counters_t` as value type
+BPF_HASH(dest_counters, struct key_t, struct counters_t, 1000);
+
+// Attach hook for the `net_dev_start_xmit` kernel trace event
+TRACEPOINT_PROBE(net, net_dev_start_xmit) {
+  struct sk_buff* skb = (struct sk_buff*) args->skbaddr;
+  struct sock* sk     = skb->sk;
+
+  struct key_t key = {};
+  bpf_probe_read(&key.daddr, sizeof(sk->sk_daddr), &sk->sk_daddr);
+  // ignore packets destined for localhost
+  // IPv4 127.0.0.1 == 16777343, 0.0.0.0 == 0
+  if ((key.daddr == 16777343 || key.daddr == 0)) {
+    return 0;
+  }
+
+  // read binary name, pid, destination IP address, port
+  bpf_get_current_comm(key.comm, TASK_COMM_LEN);
```

Comment:
This might not return the app name that triggered the traffic since the packet could be processed asynchronously in the kernel.

Reply:
True. I am seeing traffic from the swapper process with pid 0 which I think is from asynchronous processing. How do I find the app or pid that actually triggered traffic? It looks like there is no sure fire way and we have to make do with best effort.

[0] https://www.brendangregg.com/blog/2016-11-30/linux-bcc-tcplife.html. See "4. Showing task context."

Reply:
you could export the 5-tuple to the map. In userspace we could map it to pid using `ss` type of utility.
We might miss short lived sockets/process in this way, but I dont think it will matter in most of cases.

Reply:
I am curious about the choice of xmit (L2) trace point.
why not use L4 tracepoint similar to https://github.com/iovisor/bcc/blob/master/tools/tcptop.py?
This would have much lower overhead.

Reply:
We wanted to include L3 traffic from ping, traceroute etc. as well.

Reply:
do we have any guess on how much bandwidth is spent on pure L3 traffic?
is it work the overhead of tracking L4 level stats at L2 ?

Reply:
I don't have any estimates/guesses but I don't think it would be much. _Magmad_ sends some ping and traceroute traffic for monitoring periodically, but bandwidth-wise it should be miniscule. If we switch to `tcp_send`, we will also lose the packet counts and would be left with byte counts only.

Reply:
you are not getting real packet count even in this function. Most of devices has TCP optimizations that offload TCP segments to the NIC. so you are really counting TCP segments. 
Looking at the tcp_info. All required stats are already there. so the eBPF code is pretty much redundant stats collector.
We could just make use of `ss -nti`

ref: https://www.brendangregg.com/blog/2016-10-15/linux-bcc-tcptop.html

Reply:
The primary motivation for using eBPF over a ready-made approach is flexibility. Eventually, we want more than just byte counters keyed by 5-tuples.

For example, we want separate counters for traffic between pairs of AGW and orc8r services. At the network level, all this traffic may be going over a single socket (when _control_proxy_ is enabled) or over multiple sockets with the same destination host and port (when _control_proxy_ is disabled). eBPF will allow us to look into the authority HTTP header and figure out the destination service. We may also want to look at the HTTP/2 stream numbers in the future.

Another example could be open fd counters as discussed in #8173. While we currently have the `process_open_fds` metric that is collected through the `psutil` python package, that method will not work if we wanted to find out more about the kind of fds opened, how long they stayed opened etc. eBPF will allow us to instrument the `open()` and `close()` syscalls and keep track of that.

The plan is to make _kernsnoopd_ a home for all such metric collection. `packet_count` is the first such collector. The upside of this program is that it does not miss short-lived processes (assuming their traffic is not processed asynchronously,) and accounts for non-TCP traffic as well. The downside is the expected high performance impact which I'm working on measuring. I [will also write a separate program](https://github.com/magma/magma/pull/8251#discussion_r675767217) that listens to TCP events and correctly implements the AGW service to orc8r service mapping.

cc: @electronjoe

Reply:
> The primary motivation for using eBPF over a ready-made approach is flexibility. Eventually, we want more than just byte counters keyed by 5-tuples.
> 
> For example, we want separate counters for traffic between pairs of AGW and orc8r services. At the network level, all this traffic may be going over a single socket (when _control_proxy_ is enabled) or over multiple sockets with the same destination host and port (when _control_proxy_ is disabled). eBPF will allow us to look into the authority HTTP header and figure out the destination service. We may also want to look at the HTTP/2 stream numbers in the future.
> 
I see, the control proxy makes it bit complicated.
anyways, I think it make sense to use tcp level hooks to collect the packet data and expand it later.


> Another example could be open fd counters as discussed in #8173. While we currently have the `process_open_fds` metric that is collected through the `psutil` python package, that method will not work if we wanted to find out more about the kind of fds opened, how long they stayed opened etc. eBPF will allow us to instrument the `open()` and `close()` syscalls and keep track of that.
> 
> The plan is to make _kernsnoopd_ a home for all such metric collection. `packet_count` is the first such collector. The upside of this program is that it does not miss short-lived processes (assuming their traffic is not processed asynchronously,) and accounts for non-TCP traffic as well. The downside is the expected high performance impact which I'm working on measuring. I [will also write a separate program](https://github.com/magma/magma/pull/8251#discussion_r675767217) that listens to TCP events and correctly implements the AGW service to orc8r service mapping.
> 

I am  fine with the metric collection service, its good addition to AGW. I​ am bit worried about the direction this packet stats collector has taken.

Reply:
> I see, the control proxy makes it bit complicated.

Indeed. But it's complicated even without _control_proxy_ because the orc8r is listening on the same port for all cloud services.

> I​ am bit worried about the direction this packet stats collector has taken.

I understand and agree. We were worried about performance too and discussed it in the [design doc](https://magma.github.io/magma/docs/next/proposals/p018_control_network_metrics#performance). I'll change the `packet_count` program to use the `tcp_send` hook and then measure the performance impact to complete it.

Reply:
> > I see, the control proxy makes it bit complicated.
> 
> Indeed. But it's complicated even without _control_proxy_ because the orc8r is listening on the same port for all cloud services.
> 

You could make use of 5 tuples in case of direct orc8r connection. AFAICS with control proxy there is no way to figure it out without parsing the HTTP headers or mapping current pid to task (as you have done).

> > I​ am bit worried about the direction this packet stats collector has taken.
> 
> I understand and agree. We were worried about performance too and discussed it in the [design doc](https://magma.github.io/magma/docs/next/proposals/p018_control_network_metrics#performance). I'll change the `packet_count` program to use the `tcp_send` hook and then measure the performance impact to complete it.

Thanks.

Reply:
@pshelar I have now switched from instrumenting `start_xmit` to `tcp_sendmsg` and posted performance evaluation results in PR summary above.